File: 0079-word-search/0079-word-search.java
class Solution {
    public boolean exist(char[][] board, String word) {
        int m = board.length;
        int n = board[0].length;
        // boolean[][] vis = new boolean[m][n];

        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == word.charAt(0) && dfs(board, word, 0, i, j)) {
                    return true;
                }
            }
        }

        return false;
    }

    private boolean dfs(char[][] board, String word, int idx, int x, int y) {
        if (idx == word.length()) {
            return true;
        }

        int m = board.length;
        int n = board[0].length;

        if (x < 0 || x >= m || y < 0 || y >= n || board[x][y] == '*' || board[x][y] != word.charAt(idx)) {
            return false;
        }
        char c = board[x][y];
        board[x][y] = '*';

        boolean found = dfs(board, word, idx + 1, x + 1, y) ||
                        dfs(board, word, idx + 1, x - 1, y) ||
                        dfs(board, word, idx + 1, x, y + 1) ||
                        dfs(board, word, idx + 1, x, y - 1);
        // vis[x][y] = false;  
        board[x][y] = c; // backtrack

        return found;
    }
}

File: 0079-word-search/0079-word-search.cpp
class Solution {
public:
    int m,n;
    int size;
    bool dfs(vector<vector<char>>& board, int i, int j, string &word, int currIdx = 0){
        if(currIdx == size) return true;
        if(i<0 || i==m || j<0 || j==n) return false;

        if(word[currIdx] != board[i][j]) return false;
        
        char ch = board[i][j];

        board[i][j] = '$';

        bool left = dfs(board, i, j-1, word, currIdx+1);

        bool right = dfs(board, i, j+1, word, currIdx+1);

        bool top = dfs(board, i-1, j, word, currIdx+1);

        bool bottom = dfs(board, i+1, j, word, currIdx+1);

        board[i][j] = ch;

        return (left || right || top || bottom);

    }
    bool exist(vector<vector<char>>& board, string word) {
        m = board.size();
        n = board[0].size();
        size = word.size();
        for(int i=0; i<m; i++){
            for(int j=0; j<n; j++){
                if(board[i][j] == word[0]){
                    if(dfs(board, i, j, word)){
                        return true;
                    }
                }
            }
        }

        return false;
    }
};